pip/pip-281.md

```diff
@@ -0,0 +1,268 @@
+
+## In Scope
+
+<!--
+What this PIP intend to achieve once It's integrated into Pulsar.
+Why does it benefit Pulsar.
+-->
+
+- This change will be added in the extensible load balancer.
+
+## Out of Scope
+
+<!--
+Describe what you have decided to keep out of scope, perhaps left for a different PIP/s.
+-->
+
+- This won't change the existing load balancer behavior(modular load manager).
+
+
+
+# High Level Design
+
+<!--
+Describe the design of your solution in *high level*.
+Describe the solution end to end, from a birds-eye view.
+Don't go into implementation details in this section.
+
+I should be able to finish reading from beginning of the PIP to here (including) and understand the feature and 
+how you intend to solve it, end to end.
+
+DON'T
+* Avoid code snippets, unless it's essential to explain your intent.
+-->
+
+To achieve the goals above, we could modify the bundle transfer protocol by the following.
+The proposed protocol change is based on the bundle states from PIP-192. 
+
+Basically, we could close the ledgers only in the releasing state and finally disconnect clients in the owned state with destination broker urls. The clients will directly connect to the pre-assigned destination broker url without lookups.  Meanwhile, during this transfer, any produced messages will be ignored by the source broker.
+
+Current Unload and Lookup Sequence in Extensible Load Balancer
+```mermaid
+sequenceDiagram
+    participant Clients
+    participant Owner Broker
+    participant New Owner Broker
+    participant Leader Broker
+    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
+    Owner Broker ->> Owner Broker: close broker topic sessions
+    Owner Broker ->> Clients: close producers and consumers
+    Clients ->> Clients: reconnecting (inital delay 100ms)
+    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
+    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
+    Clients ->> Owner Broker: lookup
+    Owner Broker ->> Clients: redirect
+    Clients ->> New Owner Broker: lookup
+    New Owner Broker ->> Clients: return(connected)
+```
+
+Proposed Unload Sequence in Extensible Load Balancer without Lookup
+```mermaid
+sequenceDiagram
+    participant Clients
+    participant Owner Broker
+    participant New Owner Broker
+    participant Leader Broker
+    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
+    Owner Broker ->> Owner Broker: close broker topic sessions(e.g ledgers) without disconnecting producers/consumers(fenced)
+    Clients -->> Owner Broker: message pubs are ignored
+    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
+    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
+    Owner Broker ->> Owner Broker: close the fenced broker topic sessions
+    Owner Broker ->> Clients: close producers and consumers (with newOwnerBrokerUrl)
+    Clients ->> New Owner Broker: immediately connect
```

Comment:
I am wondering if there are any opportunities to decrease the latency further by making some of these steps concurrent. I spent time thinking about this last year, but I never had the chance to do anything formal. In my view, the goal should be to decouple any steps that do not actually need to be done in order as long as we can gracefully recover without introducing too much complexity. Hopefully these thoughts are helpful. :)

First, here is one of my core assumptions:

Topic transfer due to load balancing is "voluntary" (or at the very least, does not need to be immediate) so the goal should be to minimize latency as much as possible. Therefore, doing something slightly inefficient on the server side or in the protocol to minimize the observed latency is preferable.

Based on that assumption, I think the following things could help decrease transfer latency:

1. Add a state to the `ServiceUnitState` enum that allows for a target broker to claim being the "next" broker for a topic without actually becoming the owner. Then, the next broker can do things like open the next ledger for the topic without actually being the "live" owner for the topic. This could be "wasteful" in certain failure scenarios, but that extra complexity is preferable to latency spikes.
2. The new state added in step 1 would also give us the chance to tell clients to eagerly connect to the new broker before producers/consumers are disconnected so that the connection is ready to be hot swapped when the producers/consumers are closed by the broker.

In thinking about potential flow, I see two alternatives to your proposed design.

In this diagram, we would add a new protocol message that essentially tells a client to wait for the new broker url for the producer/consumer. In the event of failure (the connection closes), the client would simply do a normal lookup.

```mermaid
sequenceDiagram
    participant Clients
    participant Owner Broker
    participant New Owner Broker
    participant Leader Broker
    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
    Owner Broker ->> Owner Broker: close broker topic sessions
    Owner Broker ->> Clients: close producers and consumers (and indicate new broker command will come)
    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
    Owner Broker ->> Clients: tell clients to connect to new owner broker
    Clients ->> New Owner Broker: immediately connect
    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
    Owner Broker ->> Clients: tell clients to create producers/consumers
    Clients ->> New Owner Broker: create producers/consumers
```

This second diagram implements both points from above. I think it is probably too complicated, though I do like the idea of really trying to remove all obstacles for latency that we possibly can.

```mermaid
sequenceDiagram
    participant Clients
    participant Owner Broker
    participant New Owner Broker
    participant Leader Broker
    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
    Owner Broker ->> New Owner Broker: "state:WeaklyAssign:" weakly assign new ownership
    New Owner Broker ->> New Owner Broker: asynchronously prepare to own topic
    New Owner Broker ->> Owner Broker: "state:WeaklyOwned:" ack new weak ownership
    Owner Broker ->> Clients: tell clients to connect to new owner broker (net new protocol message)
    Clients ->> New Owner Broker: immediately connect
    Owner Broker ->> Owner Broker: close broker topic sessions
    Owner Broker ->> Clients: close producers and consumers (and indicate new broker url)
    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
    Owner Broker ->> Clients: tell clients to create producers/consumers (net new protocol command)
    Clients ->> New Owner Broker: create producers/consumers
```

Reply:
In the second diagram, perhaps it might even make more sense for the clients to tell `New Owner Broker` which topics they intend to connect to and their intended producers/consumers. That way, when the topic is loaded, the new broker can tell the producers and consumers they are created immediately upon creation.

Reply:
Thank you for sharing your thoughts.

I will read and think more about the new proposals tomorrow, but at first glance, it appears that the alternatives require brokers to send additional 'prepare' commands to clients. Don't they add complexity to the existing protocol?

Reply:
> it appears that the alternatives require brokers to send additional 'prepare' commands to clients. Don't they add complexity to the existing protocol?

Do you view adding complexity as a generic problem? The added complexity seems necessary to the problem we're solving, so I see the added commands as acceptable. If we don't send prepare commands, we get simplicity at the cost of added latency.

Reply:
We probably should consider the complexity of sending additional "prepare" commands in the protocol because the complexity could add up when there are many clients. 

Regarding this optimization, `the new broker pre-creates(async) topic ledgers.` Although I believe this parallel ledger creation can help reduce latency in most cases, in the worst case, this could cause excessive ledger creations because the dst broker needs to list all of the topics for that namespace, filter them by the bundle mapping, and create ledgers for them. We probably need to limit the number of pre-created ledgers or let the clients tell the new broker the list of topics, which will increase the code complexity.

This PIP can't optimize for all cases, but it is trying to reduce the latency with minimal complexity. I think we can push this version of the simple optimization first for this round and consider the next iterations of optimization if the community sees strong needs.

Reply:
We probably should consider the complexity of `sending additional "prepare" commands in the protocol` because the complexity could add up when there are many clients. Also, in the multi-producer/consumer client env, the connections to the new brokers could already be there; in this case, the latency optimization impact could be less.

Regarding this optimization, `the new broker pre-creates(async) topic ledgers.` Although I believe this parallel ledger creation can help reduce latency in most cases, in the worst case, this could cause excessive ledger creations because the dst broker needs to list all of the topics for that namespace, filter them by the bundle mapping, and create ledgers for them. We probably need to limit the number of pre-created ledgers or let the clients tell the new broker the list of topics, which will increase the protocol complexity and make the brokers additionally rely on clients.

This PIP can't optimize for all cases, but it is trying to reduce the latency with minimal complexity. I think we can push this version of the simple optimization first for this round and consider the next iterations of optimization if the community sees strong needs.

Reply:
Based on what you've written, I understand you think solution 2 is too complex. I agree that it is a large change and in open source, we focus on concrete improvements. As such, I would like to discuss my first solution more, which I do not think introduces nearly as much complexity. At it's heart, my first diagram is focused on removing "ignore messages" from the pulsar hand-off protocol. This is where my question on the other thread comes into play. What is the expected duration of the state transition from `Assign` to `Owned`?

Even though we likely won't go with 2, here are my in line responses to your comment:

> We probably should consider the complexity of `sending additional "prepare" commands in the protocol` because the complexity could add up when there are many clients.

Well said. I agree that increased complexity should be accompanied by appropriate analysis of the tradeoffs. And it's especially true when adding logic to the Pulsar clients, which are already quite "thick" and lacking in appropriate documentation.

> Also, in the multi-producer/consumer client env, the connections to the new brokers could already be there; in this case, the latency optimization impact could be less.

Yes, in certain configurations, there would be diminishing returns for certain attempts to eagerly hand off the connection. However, this is only one style of deployment. It is also common for many applications to have just a few producers/consumers.

> Although I believe this parallel ledger creation can help reduce latency in most cases, in the worst case, this could cause excessive ledger creations because the dst broker needs to list all of the topics for that namespace, filter them by the bundle mapping, and create ledgers for them.

I agree that we should not naively pre-create ledgers. How hard would it be for the owner broker to tell the destination broker which topics it should pre-load?

> We probably need to limit the number of pre-created ledgers or let the clients tell the new broker the list of topics, which will increase the protocol complexity and make the brokers additionally rely on clients.

This could easily work if we let the client's `PRODUCER` and `SUBSCRIBE` protocol commands take on a special meaning when the "destination broker" is in the state where it is "acquiring" the bundle but hasn't taken ownership yet. In this case, a client connects and then attempts to create a producer or a consumer. The broker takes this as the queue to start loading the topic. However, it will delay responding to the client's command until the topic is fully loaded.

Additionally, because the destination broker would not accept ownership of the topic until the broker is 99% ready to write data (there will still need to be final updates to zookeeper to add the pre-loaded ledger to the topic's list of ledgers), we minimize the risk of intermittent networking issues to zookeeper/bookkeeper because any issues that lead to a back off retry is out of the hot path, assuming the clients are not disconnected from the original broker eagerly.

> This PIP can't optimize for all cases, but it is trying to reduce the latency with minimal complexity.

I agree we must limit what we're optimizing.

Reply:
> This is where my question on the other thread comes into play. What is the expected duration of the state transition from Assign to Owned? 

It is hard to say the exact time as this time could be more or less depending on the cluster size. I observed that p99.99 pub latency was around < 2 secs in a cluster of 10 brokers with 10k topic partitions and publishers distributed to 1000 bundles at 1000 msgs/s .

> At it's heart, my first diagram is focused on removing "ignore messages" from the pulsar hand-off protocol. 

Could you explain your concerns about the "ignore messages" state? 

The "ignore messages" state is a somewhat optimistic optimization without additional commands and client-side complexity, considering 1. the transfer protocol finishes fairly quickly and 2. the ignored messages could be somewhat limited by the unloading time(also deterministically controlled by `maxPendingMessages`). During unloading, the source broker anyway cannot ack messages(ignore messages) after the ledger has been closed. When the ownership state transfer is complete, the producer will immediately resend the pending messages to the dst broker within x secs(< 2 secs in the above example). Also, the number of pending(ignored) messages can be limited by `maxPendingMessages.`

I see that if we send the `prepare` command to the clients, we can deterministically block the clients from sending messages at the cost of the additional commands to the clients and maintaining an additional `paused` state. The message latency from pending queues to the acknowledged will remain the same, but we will save some network costs by forcefully blocking message sends during unloading. 

I am worried about the complexity of this change, and the gain might be less significant, as the unloading is mostly momentary. Also, by this PIP we want to introduce minimal complexity in this round with less error-prone code.

Reply:
> It is hard to say the exact time as this time could be more or less depending on the cluster size. I observed that p99.99 pub latency was around < 2 secs in a cluster of 10 brokers with 10k topic partitions and publishers distributed to 1000 bundles at 1000 msgs/s .

This is a really valuable data point! Is the 2 seconds the time for transfer or just the time from "owner broker sends assign" to "owner broker receives owned"? Also, do you have any data comparing the current load balancing interruption to this new topic transfer protocol's interruption? It'd really help to see more data.

> Could you explain your concerns about the "ignore messages" state?

Ignoring messages is inefficient, and in my view, our goal is to build a highly optimized message broker. The owner broker is already overloaded--it is offloading a bundle--so doing something to minimize its network and memory utilization seems valuable. Further, I want to minimize the tradeoffs associated with load balancing because too many tradeoffs will lead to less aggressive load balancing. If we can avoid ignoring messages, I think we should.

It seems to me that the current load balancing hand off protocol is the naive solution, and we're looking to replace it with a more sophisticated solution. The solution we choose will be in place for a while.

> The "ignore messages" state is a somewhat optimistic optimization without additional commands and client-side complexity, considering 1. the transfer protocol finishes fairly quickly and 2. the ignored messages could be somewhat limited by the unloading time(also deterministically controlled by `maxPendingMessages`). During unloading, the source broker anyway cannot ack messages(ignore messages) after the ledger has been closed. When the ownership state transfer is complete, the producer will immediately resend the pending messages to the dst broker within x secs(< 2 secs in the above example). Also, the number of pending(ignored) messages can be limited by `maxPendingMessages.`

I agree that it will be bound by `maxPendingMessages`, but I think it is unlikely that most producers will hit their limit in single digit seconds, which indicates that the amount of unnecessary data transfer for the bundle will probably be equal to the `bytes/sec` * `duration of transfer`. I also agree that we're not really talking about ways to decrease the amount of time that messages are unacked (however, I think that eager connection to the destination broker could lead to decreased latency).

> I am worried about the complexity of this change, and the gain might be less significant, as the unloading is mostly momentary. Also, by this PIP we want to introduce minimal complexity in this round with less error-prone code.

To be clear, this is a complex change. We are replacing a naive solution that involved an abrupt disconnect followed by client polling to reconnect with a reactive solution that tells the client where to connect. The previous "naive" solution was able to avoid sending ignored messages because it was eagerly disconnected. It is fair to want to avoid overly complex solutions, but I don't think it would be surprising to add a new protocol message to solve topic transfer.

In reviewing my earlier diagram, I think we could get away with 1 new field on the close commands and 1 new protocol message. The first new field is to tell clients to disconnect without trying to reconnect immediately. The second is to tell them the new broker owner. This command optimistically assumes the assigned broker will transition to the owner broker. (Is that a safe assumption?) The broker code could be updated to make sure that when a topic (bundle) is being loaded, it doesn't reject producer and consumer create requests, but holds them on the line to prevent unnecessary retries.

```mermaid
sequenceDiagram
    participant Clients
    participant Owner Broker
    participant New Owner Broker
    participant Leader Broker
    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
    Owner Broker ->> Owner Broker: close broker topic sessions
    Owner Broker ->> Clients: close producers and consumers (and indicate new broker command will come)
    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
    Owner Broker ->> Clients: tell clients to connect to new owner broker
    Clients ->> New Owner Broker: immediately connect and attempt to create producers/consumers (broker does not reject)
    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
    New Owner Broker ->> Clients: PRODUCER_SUCCESS and SUBSCRIBE_SUCCESS commands
```

What do you think? Thanks!

Reply:
> > It is hard to say the exact time as this time could be more or less depending on the cluster size. I observed that p99.99 pub latency was around < 2 secs in a cluster of 10 brokers with 10k topic partitions and publishers distributed to 1000 bundles at 1000 msgs/s .
> 
> This is a really valuable data point! Is the 2 seconds the time for transfer or just the time from "owner broker sends assign" to "owner broker receives owned"? Also, do you have any data comparing the current load balancing interruption to this new topic transfer protocol's interruption? It'd really help to see more data.
> 

It is the p99.99 pub latency reported on the producer client. So, it should include all server-side state changes and final persist ack time.

> > Could you explain your concerns about the "ignore messages" state?
> 
> Ignoring messages is inefficient, and in my view, our goal is to build a highly optimized message broker. The owner broker is already overloaded--it is offloading a bundle--so doing something to minimize its network and memory utilization seems valuable. Further, I want to minimize the tradeoffs associated with load balancing because too many tradeoffs will lead to less aggressive load balancing. If we can avoid ignoring messages, I think we should.
> 

> It seems to me that the current load balancing hand off protocol is the naive solution, and we're looking to replace it with a more sophisticated solution. The solution we choose will be in place for a while.
> 
> > The "ignore messages" state is a somewhat optimistic optimization without additional commands and client-side complexity, considering 1. the transfer protocol finishes fairly quickly and 2. the ignored messages could be somewhat limited by the unloading time(also deterministically controlled by `maxPendingMessages`). During unloading, the source broker anyway cannot ack messages(ignore messages) after the ledger has been closed. When the ownership state transfer is complete, the producer will immediately resend the pending messages to the dst broker within x secs(< 2 secs in the above example). Also, the number of pending(ignored) messages can be limited by `maxPendingMessages.`
> 
> I agree that it will be bound by `maxPendingMessages`, but I think it is unlikely that most producers will hit their limit in single digit seconds, which indicates that the amount of unnecessary data transfer for the bundle will probably be equal to the `bytes/sec` * `duration of transfer`. I also agree that we're not really talking about ways to decrease the amount of time that messages are unacked (however, I think that eager connection to the destination broker could lead to decreased latency).
> 
> > I am worried about the complexity of this change, and the gain might be less significant, as the unloading is mostly momentary. Also, by this PIP we want to introduce minimal complexity in this round with less error-prone code.
> 
> To be clear, this is a complex change. We are replacing a naive solution that involved an abrupt disconnect followed by client polling to reconnect with a reactive solution that tells the client where to connect. The previous "naive" solution was able to avoid sending ignored messages because it was eagerly disconnected. It is fair to want to avoid overly complex solutions, but I don't think it would be surprising to add a new protocol message to solve topic transfer.
> 
> In reviewing my earlier diagram, I think we could get away with 1 new field on the close commands and 1 new protocol message. The first new field is to tell clients to disconnect without trying to reconnect immediately. The second is to tell them the new broker owner. This command optimistically assumes the assigned broker will transition to the owner broker. (Is that a safe assumption?) The broker code could be updated to make sure that when a topic (bundle) is being loaded, it doesn't reject producer and consumer create requests, but holds them on the line to prevent unnecessary retries.
> 
> ```mermaid
> sequenceDiagram
>     participant Clients
>     participant Owner Broker
>     participant New Owner Broker
>     participant Leader Broker
>     Leader Broker ->> Owner Broker: "state:Releasing:" close topic
>     Owner Broker ->> Owner Broker: close broker topic sessions
>     Owner Broker ->> Clients: close producers and consumers (and indicate new broker command will come)
>     Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
>     Owner Broker ->> Clients: tell clients to connect to new owner broker
>     Clients ->> New Owner Broker: immediately connect and attempt to create producers/consumers (broker does not reject)
>     New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
>     New Owner Broker ->> Clients: PRODUCER_SUCCESS and SUBSCRIBE_SUCCESS commands
> ```
> What do you think? Thanks!

If we want the prepare commands(e.g. PrepareCloseProducerCommand), I think we can add the dstBrokerUrls too and make the clietns immediately connect to the brokers upon `prepare`. Still any inflight messages need to be ignored at source brokers.


```mermaid
sequenceDiagram
    participant Clients
    participant Owner Broker
    participant New Owner Broker
    participant Leader Broker
    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
    Owner Broker ->> Clients: pause producers and brokers
    Clients->> New Owner Broker: open a new connection to the dest broker
    Owner Broker ->> Owner Broker: close broker topic sessions(e.g ledgers) without disconnecting producers/consumers(fenced)
    Clients -->> Owner Broker: message pubs are ignored
    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
    Owner Broker ->> Owner Broker: close the fenced broker topic sessions
    Owner Broker ->> Clients: close producers and consumers (with newOwnerBrokerUrl)
    Clients ->> New Owner Broker: connect and send create producer|consumer

```


I agree that the additional prepare command has the following advantages.
- it can forcefully pause producers sending messages, which could save network bandwidth during unloading.
- it can make clients pre-connect to the destination broker, which could help reduce the latency.

However, considering followings,

- Even if we send the "prepare" commands to the clients, any inflight messages need to be ignored anyway.
- The network bandwidth saving could be negligible: the ignored messages are limited by the maxPendingMessages(default 0 since 2.10, but previously 1000) and mem-based pending messages(64mb since 2.10). These ignored messages will also be limited by how fast the unloading finishes (Optimistically speaking, < 2 secs in the above example).
- It needs brokers to additionally send the prepare command to the clients, adding client-broker protocol complexity. 
- It adds the complexity of adding `paused` states in the clients. We have java, c++, golang code to maintain, so this will increase the scope of the work and complexity.
- The pre-connection effect can be useful only for the first send message's latency because the next send messages will reuse the connection anyway.

I believe this "prepare" command can be added in the next iteration(probably optimization phase 2), too, if we see the strong need here.

Reply:
> Still any inflight messages need to be ignored at source brokers.

I agree that there will still be a window of messages that will be received but not persisted. That is already part of the protocol.

> * It needs brokers to additionally send the prepare command to the clients, adding client-broker protocol complexity.

I really think the protocol is designed to handle this complexity. The resulting code might have multiple conditionals, but lifecycle changes are complex, and decoupling events (telling a producer to stop sending messages and telling producer which broker to connect to) seems like a more robust design, in my opinion.

Another potential issue with ignoring protocol messages from clients during unloading is that it might increase the number of duplicate messages sent to consumers because `ACK` protocol messages will be sent to the source broker but not persisted to the consumer's ledger. This is always a possibility (we work with at least once delivery after all), but our goal is to make the protocol work in a way that minimizes these kinds of dupes. The current protocol disconnects a consumer and the consumer which triggers behaviors that should decrease dupes (though I haven't looked at the source code in a while). This new protocol leaves the gap open a little longer.

My primary concerns about the pausing are in cases when the system is already degraded. It's possible that the latency is so negligible in most cases that the dupes and extra network usage are not a problem.

> * It adds the complexity of adding `paused` states in the clients. We have java, c++, golang code to maintain, so this will increase the scope of the work and complexity.

These clients already handle the "pausing" when they get disconnected from one broker and reconnect to the next. As such, I think they should all already have this state built in. There will still need to be additional logic to handle the new protocol message or protocol field, but the states should be represented there already.

> I believe this "prepare" command can be added in the next iteration(probably optimization phase 2), too, if we see the strong need here.

This is a valid perspective. My primary counterpoint is that having a phase two implementation would mean three hand-off mechanisms, which increases complexity more than switching to the "prepare" solution now.